In [2]:
#libraries to use
import pandas as pd
import numpy as np

## Table A Conversions

In [3]:
#reading data
df_conv = pd.read_csv('table_A_conversions.csv')
df_conv.head(3)

Conv_Date    Revenue                                   User_ID  \
0  06-03-2017   47.00000  5094298f068196c5349d43847de5afc9125cf989   
1  02-03-2017   98.00004                                       NaN   
2  02-03-2017  180.35300                                       NaN   

                                    Conv_ID  
0  881152bb20f9b73daafb99d77714f38ac702629c  
1  faf5c1181ea84a32237dff45ca201d2c28f19d7b  
2  b0e58a88459ece1b585ca22c93e633dc56273b83

In [4]:
#checking for missing values
df_conv.isna().sum()

Conv_Date       0
Revenue         0
User_ID      2296
Conv_ID         0
dtype: int64

**Statistical description of the revenue values**

In [99]:
df_conv.describe().transpose()

count        mean        std   min      25%       50%      75%  \
Revenue  79643.0  181.695055  109.24422  20.0  114.232  158.4706  217.592   

              max  
Revenue  4596.476

In [24]:
# convert Conv_Date to datetime format 
df_conv['Conv_Date'] = pd.to_datetime(df_conv['Conv_Date'])

In [25]:
df_conv.dtypes

Conv_Date    datetime64[ns]
Revenue             float64
User_ID              object
Conv_ID              object
dtype: object

In [43]:
# Consolidate revenue overtime sorted
df_first = df_conv.sort_values('Conv_Date', ascending=False).groupby('Conv_Date').sum()
df_first.head(10)

Revenue
Conv_Date               
2017-01-03   32614.10044
2017-01-04  308243.73544
2017-01-05   54554.81456
2017-01-06   27378.99200
2017-01-07   23815.23600
2017-01-08   24009.63200
2017-01-09   29732.62800
2017-01-10   87728.88400
2017-01-11   28170.16800
2017-01-12   20588.42000

In [42]:
df_first.to_csv('all_revenue.csv')

In [46]:
df_conv_clean = df_conv.dropna()
df_conv_clean.describe().transpose()

count        mean         std   min      25%      50%      75%  \
Revenue  77347.0  182.080439  109.673596  20.0  114.392  158.776  218.052   

              max  
Revenue  4596.476

In [48]:
n_orders = df_conv.groupby(['User_ID'])['Conv_ID'].nunique()
mult_orders_perc = np.sum(n_orders > 1) / df_conv['User_ID'].nunique()
print(f'{100 * mult_orders_perc:.2f}% of customers ordered more than once.')

16.12% of customers ordered more than once.


In [53]:
clean_conv = pd.concat(i for _, i in df_conv_clean.groupby('User_ID')).set_index('Conv_Date')
clean_conv.to_csv('clean_conv.csv')

In [30]:
#show only users that have multiple orders
pd.concat(g for _, g in df_conv.groupby("User_ID") if len(g) > 1).set_index('Conv_Date')

Revenue                                   User_ID  \
Conv_Date                                                         
2017-05-11   93.88800  0003509d64606735e66a3d32f2a1a084f613ee4b   
2017-12-28  151.74400  0003509d64606735e66a3d32f2a1a084f613ee4b   
2017-06-25  493.43200  00035f943a8a8e176fdd5a44059b38dcc0c73f5a   
2018-02-17  339.73200  00035f943a8a8e176fdd5a44059b38dcc0c73f5a   
2017-09-03  286.41152  001a50d296e7fb535c5a0cdfa3f5072528f1fd28   
...               ...                                       ...   
2018-04-02  237.56800  ffdd3feb1c2357bf148a932a80e0b279eb334b69   
2018-02-19  201.26400  ffdd3feb1c2357bf148a932a80e0b279eb334b69   
2017-06-03  148.05896  fff5a4b9d1c992c9faa86d91d77b8d4aa51a20bb   
2017-03-04  108.12000  fff5a4b9d1c992c9faa86d91d77b8d4aa51a20bb   
2017-05-22   69.14800  fff5a4b9d1c992c9faa86d91d77b8d4aa51a20bb   

                                             Conv_ID  
Conv_Date                                             
2017-05-11  374dc0ad68c61326a2edf2f7435dc47b494bb7a5  
2017-12-28  089d0519c48a6e733ced15f5cdc97d27c61fa01a  
2017-06-25  ce41e24ff8626bc311df2840352980a3dbf6937f  
2018-02-17  3ee027ad3094248776aaccaad772c79494b8f622  
2017-09-03  01c7227fd54bc1bd318c55c4d15073d843dabef9  
...                                              ...  
2018-04-02  15dcfd074f41a176098f47d937e432845785a0f5  
2018-02-19  a6d1bcdcfb50952ebc7b49ed501e2c798efc5a22  
2017-06-03  d2fe4e36e8c2ed270c16f30639cb6d486d760362  
2017-03-04  0c031f194131b8fcda3a98cc64ad4220c14f6bfd  
2017-05-22  99d9db942d7b1043797c3136e58019bd0b1e1b1c  

[30915 rows x 3 columns]

In [33]:
users = pd.concat(g for _, g in df_conv.groupby("User_ID") if len(g) > 1).set_index('Conv_Date')
users.to_csv('cust.csv')

In [68]:
num_orders = users.groupby(['User_ID'])['Conv_ID'].nunique().sort_values(ascending=False)
num_orders.to_csv('num_orders.csv')
num_orders.head(10)

User_ID
2c75940486d75040f269c9671ab746dffefe9692    111
8a1846b853d9522214daba775a46789ada386c3a     41
31e3c730764f2913e56fcae325f92a82bc94a4aa     30
a00698b0fc7532f015e0b0bc14d9f9b06f04fac2     29
310e3421aa1d5ff61b48cc153b460123218c0d10     28
0ad05472146efb8b505f113c4cdc3a88b5a89f41     27
be7cd84e8b175933f5b86276b429d87d414b5f4a     25
f367fc9a7cd5b3351f4741fa2b3e99f471dfd359     23
72df33e2b3ccfebff04123e211ef07d5f39a2324     23
eef3db4f1fda1eb4125fa2ec16a503b915c9a827     23
Name: Conv_ID, dtype: int64

## Table B Attribution

In [7]:
df_attr = pd.read_csv('table_b_attribution.csv')
df_attr.head(3)

Channel  IHC_Conv                                   Conv_ID
0       H  1.000000  881152bb20f9b73daafb99d77714f38ac702629c
1       I  0.300250  faf5c1181ea84a32237dff45ca201d2c28f19d7b
2       A  0.322839  faf5c1181ea84a32237dff45ca201d2c28f19d7b

In [8]:
#checking for missing values
df_attr.isna().sum()

Channel     0
IHC_Conv    0
Conv_ID     0
dtype: int64

In [90]:
attr = pd.merge(df_conv, df_attr, how='right', on='Conv_ID')
attr.head()

Conv_Date    Revenue                                   User_ID  \
0 2017-06-03   47.00000  5094298f068196c5349d43847de5afc9125cf989   
1 2017-02-03   98.00004                                       NaN   
2 2017-02-03   98.00004                                       NaN   
3 2017-02-03   98.00004                                       NaN   
4 2017-02-03  180.35300                                       NaN   

                                    Conv_ID Channel  IHC_Conv  
0  881152bb20f9b73daafb99d77714f38ac702629c       H  1.000000  
1  faf5c1181ea84a32237dff45ca201d2c28f19d7b       I  0.300250  
2  faf5c1181ea84a32237dff45ca201d2c28f19d7b       A  0.322839  
3  faf5c1181ea84a32237dff45ca201d2c28f19d7b       E  0.376911  
4  b0e58a88459ece1b585ca22c93e633dc56273b83       H  1.000000

In [94]:
attr['Conv_Date'] = pd.to_datetime(attr['Conv_Date'])
attr['Channel_val'] = attr['Revenue'] * attr['IHC_Conv']
attr = attr.set_index('Conv_Date')


In [98]:
attr.groupby('Channel')['Revenue'].sum()

Channel
A    7.909567e+06
B    4.248432e+06
C    1.679483e+06
D    4.752207e+05
E    2.950673e+06
F    2.839660e+05
G    6.782845e+06
H    5.350637e+06
I    4.436182e+06
J    1.017727e+06
K    1.026579e+06
L    7.263449e+05
M    8.028850e+05
N    3.042530e+05
O    5.677919e+04
P    9.632102e+04
Q    5.973124e+03
R    1.427021e+05
S    2.736319e+05
T    1.842404e+04
U    2.339314e+04
V    8.090468e+03
Name: Revenue, dtype: float64

## Cohort Analysis

In [100]:
import datetime as dt

In [109]:
def get_month(x):
    return dt.datetime(x.year, x.month, 1)

In [110]:
df_conv_clean['Conv_Month'] = df_conv_clean['Conv_Date'].apply(get_month)

In [112]:
df_conv_clean['Cohort_Month'] = df_conv_clean.groupby('User_ID')['Conv_Month'].transform('min')

In [114]:
df_conv_clean.drop(['Cohort'], axis='columns', inplace=True)

In [115]:
df_conv_clean.head()

User_ID  \
0   5094298f068196c5349d43847de5afc9125cf989   
3   433fdf385e33176cf9b0d67ecf383aa928fa261c   
9   59853624ad85789cc5c865405e741c612da334da   
12  8b14ce0002a2a7adbe61c6386dd5b8b8a8435222   
13  b7d38ff2747f3ebe3c6b4e6169dbb7c761177a2e   

                                     Conv_ID  Conv_Date Conv_Month  \
0   881152bb20f9b73daafb99d77714f38ac702629c 2017-06-03 2017-06-01   
3   f0e6b7de22332c7b18c024e550bb1d860130cdf1 2017-03-23 2017-03-01   
9   3687a104d8ffdfe17f7db7b08a200d4d6d184c26 2017-03-31 2017-03-01   
12  643cc4e0b0baa259d91419bbc5b535002d86e338 2017-03-14 2017-03-01   
13  7ebdca036560070819aa40c4ad8e8e66df2e5bc5 2017-10-03 2017-10-01   

   Cohort_Month  
0    2017-01-01  
3    2017-01-01  
9    2017-01-01  
12   2017-03-01  
13   2017-10-01

In [116]:
def get_date(df, column):
    year = df[column].dt.year
    month = df[column].dt.month
    day = df[column].dt.day
    return year, month, day

In [117]:
conv_year, conv_month, _ = get_date(df_conv_clean, 'Conv_Month')
cohort_year, cohort_month, _ = get_date(df_conv_clean, 'Cohort_Month')

In [118]:
year_diff = conv_year - cohort_year
month_diff = conv_month - cohort_month

In [119]:
df_conv_clean['Cohort_Index'] = year_diff * 12 + month_diff + 1

In [122]:
cohort_data = df_conv_clean.groupby(['Cohort_Month', 'Cohort_Index'])['User_ID'].apply(pd.Series.nunique).reset_index()

In [123]:
cohort_count = cohort_data.pivot_table(index = 'Cohort_Month', columns = 'Cohort_Index', values = 'User_ID')

In [124]:
cohort_count

Cohort_Index      1      2      3      4      5      6      7      8      9   \
Cohort_Month                                                                   
2017-01-01    2773.0  123.0  258.0  220.0  246.0  303.0  338.0  302.0  319.0   
2017-02-01    2014.0  206.0  128.0  153.0  179.0  202.0  168.0  204.0  131.0   
2017-03-01    4689.0  403.0  398.0  447.0  499.0  426.0  429.0  310.0  398.0   
2017-04-01    3152.0  194.0  238.0  201.0  232.0  217.0  182.0  212.0  149.0   
2017-05-01    3536.0  195.0  236.0  218.0  204.0  160.0  212.0  144.0   81.0   
2017-06-01    4017.0  235.0  202.0  178.0  133.0  198.0  104.0   83.0   80.0   
2017-07-01    4604.0  191.0  174.0  128.0  199.0   86.0   68.0   62.0   37.0   
2017-08-01    3958.0  143.0  115.0  152.0   79.0   57.0   66.0   30.0    6.0   
2017-09-01    4044.0   80.0  123.0   65.0   48.0   48.0   22.0    5.0    7.0   
2017-10-01    3218.0   86.0   50.0   26.0   36.0   18.0    3.0    6.0    4.0   
2017-11-01    4708.0   58.0   31.0   32.0   23.0    4.0    4.0    8.0    2.0   
2017-12-01    2752.0   32.0   38.0   13.0    3.0    3.0    5.0    1.0   18.0   
2018-01-01    2858.0   59.0   21.0    8.0   13.0    7.0    4.0   22.0    7.0   
2018-02-01    3034.0   34.0    7.0    8.0    4.0    5.0   26.0   12.0    4.0   
2018-03-01    2026.0    1.0    5.0    2.0    2.0    6.0    1.0    1.0    2.0   
2018-04-01     453.0    NaN    NaN    NaN    1.0    2.0    NaN    NaN    NaN   
2018-05-01     415.0    NaN    1.0    2.0    NaN    NaN    1.0    1.0    NaN   
2018-06-01     368.0    1.0    3.0    NaN    NaN    NaN    NaN    NaN    NaN   
2018-07-01     343.0    1.0    NaN    NaN    NaN    1.0    NaN    NaN    NaN   
2018-08-01     952.0    2.0    1.0    NaN    2.0    NaN    NaN    NaN    NaN   
2018-09-01     468.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2018-10-01     274.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2018-11-01     350.0    1.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2018-12-01     346.0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

Cohort_Index     10  ...    15    16    17    18    19    20    21    22  \
Cohort_Month         ...                                                   
2017-01-01    195.0  ...  54.0  25.0  21.0  19.0  16.0  83.0  27.0  12.0   
2017-02-01    210.0  ...   8.0  12.0  15.0  12.0  50.0  19.0   6.0  13.0   
2017-03-01    236.0  ...  33.0  14.0  28.0  82.0  41.0  16.0  14.0  17.0   
2017-04-01    128.0  ...  12.0   9.0  47.0  20.0   9.0  10.0  12.0   NaN   
2017-05-01     80.0  ...   7.0  51.0  14.0   8.0  11.0   6.0   NaN   NaN   
2017-06-01     30.0  ...  36.0  11.0  12.0   9.0   7.0   NaN   NaN   NaN   
2017-07-01     11.0  ...   6.0   8.0   7.0   8.0   NaN   NaN   NaN   NaN   
2017-08-01     14.0  ...   6.0   6.0   4.0   NaN   NaN   NaN   NaN   NaN   
2017-09-01      3.0  ...   4.0   7.0   NaN   NaN   NaN   NaN   NaN   NaN   
2017-10-01      5.0  ...   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2017-11-01     29.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2017-12-01      5.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-01-01      4.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-02-01      6.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-03-01      2.0  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-04-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-05-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-06-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-07-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-08-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-09-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-10-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-11-01      NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
2018-12-01      NaN

## Summary

From the analysis on the conversion and attribution data we have found that the revenue for the year 2008 had a decrease slop as with some peaks around the mid of the year. And as the year went by the revenue level was stable, There are some variables to look at to describe this behavior.

From the data that include User_ID we can se that only 16.12% of the user have more than one conversion. Ann most of these recurring customers placed two orders between 2017 and 2018.
We can see that there are four predominant marketing channels that have most of the share revenue, A, G, H, B. They have about 77% of the revenue.

The cohort analysis show us the same trend that the revenue was showing us in the first graph, as time goes by the number of active customers decrease.